### PR TITLE
Systemify

### PR DESCRIPTION
**Pull request by thebluefish**
WIP release notes
---

## bevy_commandify 0.3.0

[github](<https://github.com/thebluefish/bevy_commandify>) | [crates.io](<https://crates.io/crates/bevy_commandify>)

- Supports systems as commands, ie:
```rust
#[command]
fn foo(In(n): In<usize>, mut m: ResMut<TestUsize>) {
    **m -= n;
}

#[command]
fn irony(mut commands: Commands) -> &mut Self {
    commands.foo(5);
}
```
- Original system functions are now kept intact, sans special return type
- Bumped bevy support to `0.13`
- Supports `mut` parameters
- Properly propagates doc comments to trait methods

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -18,6 +18,12 @@ fn foo(world: &mut World, n: usize) {
     **bar -= n;
 }
 
+/// Regular bevy systems can be used as commands, too
```
Comment (spectria-limina) [inline README.md:21]: foo isn't a regular Bevy system

File: src/gen.rs [hunk 2]
```diff
@@ -19,7 +19,7 @@ pub(crate) fn commandify(
         vis,
         sig,
         block,
-    } = item;
+    } = item.clone();
     let Signature {
```
Comment (spectria-limina) [inline src/gen.rs]: You can merge this pattern with the previous one.

File: src/gen.rs [hunk 3]
```diff
@@ -113,6 +115,7 @@ pub(crate) fn commandify(
     ));
     let ecs_root = ecs_root.unwrap_or(parse_quote!(::bevy::ecs));
```
Comment (spectria-limina) [inline src/gen.rs]: This is drive-by, but all these `unwrap_or` should be `unwrap_or_else` to avoid pointlessly formatting strings

File: src/gen.rs [hunk 4]
```diff
@@ -113,6 +115,7 @@ pub(crate) fn commandify(
...
 
+    // parse generics
```
Comment (spectria-limina) [inline src/gen.rs:73]: TODO to split this out to parse module too?

File: src/gen.rs [hunk 5]
```diff
@@ -137,132 +140,273 @@ pub(crate) fn commandify(
...
+                asyncness,
+                unsafety,
+                abi,
+                fn_token,
+                ident,
+                generics,
+                inputs,
+                variadic,
+                ..
+            } = sig;
```
Comment (spectria-limina) [inline src/gen.rs]: Why are we destructuring the original `ItemFn` again?

File: src/gen.rs [hunk 6]
```diff
@@ -137,132 +140,273 @@ pub(crate) fn commandify(
...
         ));
     }
 
+    // generate fragments to be combined later
+
+    // break apart the function and piece it back together sans return type
+    // since we handle that specially above
+    let fn_frag = match &args {
+        SystemArgs::Exclusive { .. } => {
+            quote!()
```
Comment (spectria-limina) [inline src/gen.rs]: I think some explanation of why this is empty in this case would be helpful.

File: src/gen.rs [hunk 7]
```diff
@@ -137,132 +140,273 @@ pub(crate) fn commandify(
...
     } else {
-        quote!( { #(#struct_fields)* } )
+        quote!( { #(pub #fields,)* } )
     };
 
-    let apply_params = if entity_command {
-        quote!((self, #entity_field, #world_field))
-    } else {
-        quote!((self, #world_field))
+    let system_in_frag = match &args {
```
Comment (spectria-limina) [inline src/gen.rs:154]: I think each of these big fragments in the next few sections should probably be helper functions if at all possible. It's hard to follow the code particularly in a big function.

It would also help to have comments explaining what each fragment is actually generating, I think, like an example of it.

File: src/parse.rs [hunk 8]
```diff
@@ -1,9 +1,234 @@
...
+
+                exclusive_fields.push(quote!(#name: #ty));
+                exclusive_def_field_names.push(quote!(#name));
+                exclusive_impl_field_names.push(quote!(#pat));
+            }
+        }
+    }
+
+    // figure these out late since some parts have different meanings depending on whether this is an exclusive or normal system
+    let fields = if world_field.is_some() {
```
Comment (spectria-limina) [inline src/parse.rs]:
```rust
let (fields, def_field_names, impl_field names) = if world_field.is_some() {
    ...
} else {
    ...
}

File: README.md (hunk 1 above)
Comment (thebluefish) [reply README.md:21]: This comment is meant to highlight that `irony` is a normal bevy system being transformed into a Command.
